File: python/src/runner.py
from typing import Iterable
from .algorithms.base import SortAlgorithm
from .dataset import random_list
from .metrics import Metrics
import time


def run(
        sorters: Iterable[type[SortAlgorithm]],
        sizes=(1_000, 100_000, 1_000_000)
):
    results = []

    for n in sizes:
        data = random_list(n)

        for sorter_cls in sorters:
            sorter = sorter_cls()
            data_copy = data[:]
            m = Metrics()
            start = time.perf_counter_ns()
            sorter.sort(data_copy)
            m.elapsed_ns = time.perf_counter_ns() - start
            results.append(
                {
                    "algorithm": sorter.name,
                    "N": n,
                    "time_ns": m.elapsed_ns
                }
            )

    return results


def run_one(sorter: SortAlgorithm, n: int) -> dict:
    data = random_list(n)
    start = time.perf_counter()
    m = Metrics()
    sorter.sort(data, m)
    elapsed = time.perf_counter() - start


    return {
        "algorithm": sorter.name,
        "n": n,
        "time_s": elapsed,
    }

File: rust/src/runner.rs
use std::time::Instant;

use crate::{
    algorithms::base::SortAlgorithm,
    dataset,
    metrics::Metrics,
};

#[derive(serde::Serialize)]
pub struct Record {
    pub algorithm: String,
    pub n: usize,
    pub elapsed_ns: u128,
}

pub fn run(
    algorithms: &[Box<dyn SortAlgorithm>],
    sizes: &[usize],
) -> Vec<Record> {
    let mut results: Vec<Record> = Vec::new();

    for &n in sizes {
        let base_data = dataset::random_vec(n);
        for algo in algorithms {
            let mut data = base_data.clone();
            let mut m = Metrics::new();

            let start = Instant::now();
            algo.sort(&mut data);

            m.elapsed = start.elapsed();

            results.push(Record {
                algorithm: algo.name().into(),
                n,
                elapsed_ns: m.elapsed.as_nanos(),
            });
        }
    }

    results
}

pub fn run_one(algorithm: &dyn SortAlgorithm, n: usize) -> std::io::Result<Record> {
    use std::time::Instant;
    use crate::dataset::random_vec;

    let mut data = random_vec(n);
    let start = Instant::now();
    algorithm.sort(&mut data);
    Ok(Record {
        algorithm: algorithm.name().into(),
        n,
        elapsed_ns: start.elapsed().as_nanos(),
    })
}
